File: src/test/java/de/pflugradts/pwman3/domain/model/transfer/BytesTest.java
package de.pflugradts.pwman3.domain.model.transfer;

import java.util.Collections;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.IntStream;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;

class BytesTest {

    @Test
    void shouldHasCorrectSize() {
        // given
        final var givenBytes = new byte[]{1, 2, 3};
        final var bytes = Bytes.of(givenBytes);
        final var expectedSize = 3;

        // when
        final var actual = bytes.size();

        // then
        assertThat(actual).isNotNull().isEqualTo(expectedSize);
    }

    @Test
    void shouldGetByte() {
        // given
        final var givenBytes = new byte[]{1, 2, 3};
        final var bytes = Bytes.of(givenBytes);

        // when / then
        assertThat(givenBytes.length).isEqualTo(3);
        for (int i = 0; i < givenBytes.length; i++) {
            assertThat(bytes.getByte(i)).isEqualTo(givenBytes[i]);
        }
    }

    @Test
    void shouldGetByte_ThrowArrayIndexOutOfBoundsException_OnInvalidIndex() {
        // given
        final var givenBytes = new byte[]{1, 2, 3};
        final var invalidIndex = 4;
        final var bytes = Bytes.of(givenBytes);

        // when
        final Throwable throwable = catchThrowable(() -> bytes.getByte(invalidIndex));

        // then
        assertThat(throwable).isNotNull().isInstanceOf(ArrayIndexOutOfBoundsException.class);
    }

    @Test
    void shouldSlice() {
        // given
        final var givenBytes = new byte[]{1, 2, 3, 4};
        final var bytes = Bytes.of(givenBytes);

        // when
        final var actual = bytes.slice(1, 3);

        // then
        assertThat(actual).isNotNull().containsExactly(givenBytes[1], givenBytes[2]);
    }

    @Test
    void shouldSlice_ReturnEmptyBytes_OnInvalidRange() {
        // given
        final var givenBytes = new byte[]{1, 2, 3, 4};
        final var bytes = Bytes.of(givenBytes);

        // when
        final var actual = bytes.slice(3, 1);

        // then
        assertThat(actual).isNotNull().isEmpty();
    }

    @Test
    void shouldSlice_ThrowArrayIndexOutOfBoundsException_OnInvalidIndex() {
        // given
        final var givenBytes = new byte[]{1, 2, 3};
        final var invalidIndex = 4;
        final var bytes = Bytes.of(givenBytes);

        // when
        final Throwable throwable = catchThrowable(() -> bytes.slice(invalidIndex, invalidIndex + 1));

        // then
        assertThat(throwable).isNotNull().isInstanceOf(ArrayIndexOutOfBoundsException.class);
    }

    @Test
    void shouldScrambleAllBytes() {
        // given
        final var minAsciiValue = 32;
        final var maxAsciiValue = 126;
        final var size = 100;
        final var byteList = Collections.nCopies(size,
                Byte.valueOf(String.valueOf(maxAsciiValue + 1)));
        final var bytes = Bytes.of(byteList);

        // when
        bytes.forEach(b -> assertThat(b).isEqualTo((byte) (maxAsciiValue + 1)));
        bytes.scramble();

        // then
        assertThat(bytes.size()).isEqualTo(size);
        bytes.forEach(b -> assertThat(b).isBetween((byte) minAsciiValue, (byte) maxAsciiValue));
    }

    @Nested
    class InstantiationTest {

        @Test
        void shouldInstantiateFromByteArray() {
            // given
            final var givenBytes = new byte[]{1, 2, 3};
            final var expectedBytes = new Byte[]{givenBytes[0], givenBytes[1], givenBytes[2]};

            // when
            final var actual = Bytes.of(givenBytes);

            // then
            assertThat(actual).isNotNull().containsExactly(expectedBytes);
        }

        @Test
        void shouldInstantiateFromCharArray() {
            // given
            final var givenChars = new char[]{'a', 'b', 'c'};
            final var expectedBytes = new Byte[]{(byte) givenChars[0], (byte) givenChars[1], (byte) givenChars[2]};

            // when
            final var actual = Chars.of(givenChars).toBytes();

            // then
            assertThat(actual).isNotNull().containsExactly(expectedBytes);
        }

        @Test
        void shouldInstantiateFromByteList() {
            // given
            final var givenBytes = List.of(Byte.valueOf("1"), Byte.valueOf("2"), Byte.valueOf("3"));
            final var expectedBytes = new Byte[]{givenBytes.get(0), givenBytes.get(1), givenBytes.get(2)};

            // when
            final var actual = Bytes.of(givenBytes);

            // then
            assertThat(actual).isNotNull().containsExactly(expectedBytes);
        }

        @Test
        void shouldInstantiateFromString() {
            // given
            final var givenString = "hello";
            final var primitiveBytes = givenString.getBytes();
            final var expectedBytes = IntStream.range(0, primitiveBytes.length)
                    .mapToObj(i -> primitiveBytes[i])
                    .toArray(Byte[]::new);

            // when
            final var actual = Bytes.of(givenString);

            // then
            assertThat(actual).isNotNull().containsExactly(expectedBytes);
        }

        @Test
        void shouldInstantiateEmptyBytes() {
            // given / when / then
            assertThat(Bytes.empty()).isNotNull().isEmpty();
        }

        @Test
        void shouldCloneConstructorInput() {
            // given
            final var givenBytes = new byte[]{1, 2, 3};
            final var expectedBytes = new Byte[]{1, 2, 3};

            // when
            final var actual = Bytes.of(givenBytes);
            givenBytes[0] = 4; // change original array

            // then
            assertThat(actual).isNotNull().containsExactly(expectedBytes);
            assertThat(givenBytes).isEqualTo(new byte[]{4, 2, 3});
        }

    }

    @Nested
    class TransformationTest {

        @Test
        void shouldTransformToByteArray() {
            // given
            final var givenBytes = new byte[]{1, 2, 3};
            final var bytes = Bytes.of(givenBytes);

            // when
            final var actual = bytes.toByteArray();

            // then
            assertThat(actual).isNotNull()
                    .isEqualTo(givenBytes)
                    .isNotSameAs(givenBytes);
        }

        @Test
        void shouldTransformToCharArray() {
            // given
            final var givenChars = new char[]{'a', 'b', 'c'};
            final var referenceChars = new char[]{'a', 'b', 'c'};
            final var bytes = Chars.of(givenChars).toBytes();

            // when
            final var actual = bytes.toChars();

            // then
            assertThat(actual).isNotNull().extracting(Chars::toCharArray).isEqualTo(referenceChars);
        }

        @Test
        void shouldTransformToString() {
            // given
            final var givenString = "hello";
            final var bytes = Bytes.of(givenString);

            // when
            final var actual = bytes.asString();

            // then
            assertThat(actual).isNotNull().isEqualTo(givenString);
        }

        @Test
        void shouldCloneBytesOnCopy() {
            // given
            final var givenBytes = new byte[]{1, 2, 3};
            final var bytes = Bytes.of(givenBytes);

            // when
            final var clonedBytes = bytes.copy();
            final var actual = bytes.copy();

            // then
            assertThat(actual).isNotNull()
                    .isEqualTo(clonedBytes)
                    .isNotSameAs(clonedBytes);
        }

    }

    @Nested
    class IteratorTest {

        @Test
        void shouldIterateOverElements() {
            // given
            final var givenBytes = new byte[]{1, 2, 3};
            final var actualBytes = new byte[givenBytes.length];
            final var iterator = Bytes.of(givenBytes).iterator();

            // when
            var index = 0;
            while (iterator.hasNext()) {
                actualBytes[index++] = iterator.next();
            }

            // then
            assertThat(actualBytes).isEqualTo(givenBytes);
        }

        @Test
        void shouldThrowNoSuchElementException_OnNext_WhenThereIsNoMoreElement() {
            // given
            final var givenBytes = new byte[]{1, 2, 3};
            final var iterator = Bytes.of(givenBytes).iterator();
            while (iterator.hasNext()) {
                iterator.next();
            }

            // when
            final Throwable throwable = catchThrowable(iterator::next);

            // then
            assertThat(throwable).isNotNull().isInstanceOf(NoSuchElementException.class);
        }

    }

}

File: src/test/kotlin/de/pflugradts/passbird/domain/model/transfer/BytesTest.kt
package de.pflugradts.passbird.domain.model.transfer

import de.pflugradts.kotlinextensions.tryCatching
import de.pflugradts.passbird.domain.model.transfer.Bytes.Companion.bytesOf
import de.pflugradts.passbird.domain.model.transfer.Bytes.Companion.emptyBytes
import de.pflugradts.passbird.domain.model.transfer.Chars.Companion.charsOf
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import strikt.api.expectThat
import strikt.assertions.containsExactly
import strikt.assertions.isA
import strikt.assertions.isEmpty
import strikt.assertions.isEqualTo
import strikt.assertions.isGreaterThanOrEqualTo
import strikt.assertions.isLessThanOrEqualTo
import strikt.assertions.isNotSameInstanceAs
import strikt.assertions.isTrue
import java.util.Collections

class BytesTest {

    @Test
    fun test() {
        val givenBytes = byteArrayOf(1, 2, 3)
        val bytes = bytesOf(givenBytes)
        println(bytes.toByteArray().toList())
        println(bytes.toMutableList().also { it.remove(1.toByte()) })
    }

    @Test
    fun `should have correct size`() {
        // given
        val givenBytes = byteArrayOf(1, 2, 3)
        val bytes = bytesOf(givenBytes)
        val expectedSize = 3

        // when
        val actual = bytes.size

        // then
        expectThat(actual) isEqualTo expectedSize
    }

    @Test
    fun `should get byte`() {
        // given
        val givenBytes = byteArrayOf(1, 2, 3)
        val bytes = bytesOf(givenBytes)

        // when / then
        expectThat(givenBytes.size) isEqualTo 3
        givenBytes.indices.forEach {
            expectThat(bytes.getByte(it)) isEqualTo givenBytes[it]
        }
    }

    @Test
    fun `should throw exception on invalid index`() {
        // given
        val givenBytes = byteArrayOf(1, 2, 3)
        val invalidIndex = 4
        val bytes = bytesOf(givenBytes)

        // when
        val exception = tryCatching { bytes.getByte(invalidIndex) }.exceptionOrNull()

        // then
        expectThat(exception).isA<ArrayIndexOutOfBoundsException>()
    }

    @Test
    fun `should return slice`() {
        // given
        val givenBytes = byteArrayOf(1, 2, 3, 4)
        val bytes = bytesOf(givenBytes)

        // when
        val actual = bytes.slice(1, 3)

        // then
        expectThat(actual).containsExactly(givenBytes[1], givenBytes[2])
    }

    @Test
    fun `should return empty bytes on negative range`() {
        // given
        val givenBytes = byteArrayOf(1, 2, 3, 4)
        val bytes = bytesOf(givenBytes)

        // when
        val actual = bytes.slice(3, 1)

        // then
        expectThat(actual.isEmpty).isTrue()
    }

    @Test
    fun `should throw exception on out of bounds range`() {
        // given
        val givenBytes = byteArrayOf(1, 2, 3)
        val invalidIndex = 4
        val bytes = bytesOf(givenBytes)

        // when
        val exception = tryCatching { bytes.slice(invalidIndex, invalidIndex + 1) }.exceptionOrNull()

        // then
        expectThat(exception).isA<ArrayIndexOutOfBoundsException>()
    }

    @Test
    fun `should scramble all bytes`() {
        // given
        val minAsciiValue = 32
        val maxAsciiValue = 126
        val size = 100
        val byteList = Collections.nCopies(size, (maxAsciiValue + 1).toString().toByte())
        val bytes = bytesOf(byteList)

        // when
        bytes.forEach {
            expectThat(it) isEqualTo (maxAsciiValue + 1).toByte()
        }
        bytes.scramble()

        // then
        expectThat(bytes.size) isEqualTo size
        bytes.forEach {
            expectThat(it) isGreaterThanOrEqualTo minAsciiValue.toByte() isLessThanOrEqualTo maxAsciiValue.toByte()
        }
    }

    @Nested
    inner class InstantiationTest {
        @Test
        fun `should instantiate from byte array`() {
            // given
            val givenBytes = byteArrayOf(1, 2, 3)
            val expectedBytes = byteArrayOf(givenBytes[0], givenBytes[1], givenBytes[2])

            // when
            val actual = bytesOf(givenBytes)

            // then
            expectThat(actual).containsExactly(*expectedBytes.toTypedArray())
        }

        @Test
        fun `should instantiate from char array`() {
            // given
            val givenChars = charArrayOf('a', 'b', 'c')
            val expectedBytes = byteArrayOf(givenChars[0].code.toByte(), givenChars[1].code.toByte(), givenChars[2].code.toByte())

            // when
            val actual = charsOf(givenChars).toBytes()

            // then
            expectThat(actual).containsExactly(*expectedBytes.toTypedArray())
        }

        @Test
        fun `should instantiate from byte list`() {
            // given
            val givenBytes = listOf("1".toByte(), "2".toByte(), "3".toByte())
            val expectedBytes = byteArrayOf(givenBytes[0], givenBytes[1], givenBytes[2])

            // when
            val actual = bytesOf(givenBytes)

            // then
            expectThat(actual).containsExactly(*expectedBytes.toTypedArray())
        }

        @Test
        fun `should instantiate from string`() {
            // given
            val givenString = "hello"
            val primitiveBytes = givenString.toByteArray()
            val expectedBytes = primitiveBytes.copyOf()

            // when
            val actual = bytesOf(givenString)

            // then
            expectThat(actual).containsExactly(*expectedBytes.toTypedArray())
        }

        @Test
        fun `should instantiate empty bytes`() {
            // given / when / then
            expectThat(emptyBytes().isEmpty).isTrue()
        }

        @Test
        fun `should clone constructor input`() {
            // given
            val givenBytes = byteArrayOf(1, 2, 3)
            val expectedBytes = byteArrayOf(1, 2, 3)

            // when
            val actual = bytesOf(givenBytes)
            givenBytes[0] = 4 // change original array

            // then
            expectThat(actual).containsExactly(*expectedBytes.toTypedArray())
            expectThat(givenBytes) isEqualTo byteArrayOf(4, 2, 3)
        }
    }

    @Nested
    inner class InstantiationFromEmptySourceTest {
        @Test
        fun `should instantiate from empty byte array`() {
            // given
            val givenBytes = byteArrayOf()
            val expectedBytes = emptyBytes()

            // when
            val actual = bytesOf(givenBytes)

            // then
            expectThat(actual).containsExactly(*expectedBytes.toByteArray().toTypedArray())
        }

        @Test
        fun `should instantiate from empty char array`() {
            // given
            val givenChars = charArrayOf()
            val expectedBytes = emptyBytes()

            // when
            val actual = charsOf(givenChars).toBytes()

            // then
            expectThat(actual).containsExactly(*expectedBytes.toByteArray().toTypedArray())
        }

        @Test
        fun `should instantiate from empty byte list`() {
            // given
            val givenBytes = emptyList<Byte>()
            val expectedBytes = emptyBytes()

            // when
            val actual = bytesOf(givenBytes)

            // then
            expectThat(actual).containsExactly(*expectedBytes.toByteArray().toTypedArray())
        }

        @Test
        fun `should instantiate from empty string`() {
            // given
            val givenString = ""
            val expectedBytes = emptyBytes()

            // when
            val actual = bytesOf(givenString)

            // then
            expectThat(actual).containsExactly(*expectedBytes.toByteArray().toTypedArray())
        }
    }

    @Nested
    inner class TransformationTest {
        @Test
        fun `should transform to byte array`() {
            // given
            val givenBytes = byteArrayOf(1, 2, 3)
            val bytes = bytesOf(givenBytes)

            // when
            val actual = bytes.toByteArray()

            // then
            expectThat(actual) isEqualTo givenBytes isNotSameInstanceAs givenBytes
        }

        @Test
        fun `should transform to char array`() {
            // given
            val givenChars = charArrayOf('a', 'b', 'c')
            val referenceChars = charArrayOf('a', 'b', 'c')
            val bytes = charsOf(givenChars).toBytes()

            // when
            val actual = bytes.toChars()

            // then
            expectThat(actual.toCharArray()) isEqualTo referenceChars
        }

        @Test
        fun `should transform to string`() {
            // given
            val givenString = "hello"
            val bytes = bytesOf(givenString)

            // when
            val actual = bytes.asString()

            // then
            expectThat(actual).isEqualTo(givenString)
        }

        @Test
        fun `should clone bytes on copy`() {
            // given
            val givenBytes = byteArrayOf(1, 2, 3)
            val bytes = bytesOf(givenBytes)

            // when
            val clonedBytes = bytes.copy()
            val actual = bytes.copy()

            // then
            expectThat(actual) isEqualTo clonedBytes isNotSameInstanceAs clonedBytes
        }

        @Test
        fun `should transform to stream`() {
            // given
            val givenBytes = bytesOf("myBytes")

            // when
            val actual = givenBytes.stream()

            // then
            expectThat(actual.toList()).containsExactly(
                'm'.code.toByte(),
                'y'.code.toByte(),
                'B'.code.toByte(),
                'y'.code.toByte(),
                't'.code.toByte(),
                'e'.code.toByte(),
                's'.code.toByte(),
            )
        }
    }

    @Nested
    inner class TransformEmptyBytesTest {
        @Test
        fun `should transform to byte array`() {
            // given
            val bytes = emptyBytes()

            // when
            val actual = bytes.toByteArray()

            // then
            expectThat(actual).isEmpty()
        }

        @Test
        fun `should transform to char array`() {
            // given
            val bytes = emptyBytes()

            // when
            val actual = bytes.toChars()

            // then
            expectThat(actual.toCharArray()) isEqualTo charArrayOf()
        }

        @Test
        fun `should transform to string`() {
            // given
            val bytes = emptyBytes()

            // when
            val actual = bytes.asString()

            // then
            expectThat(actual).isEmpty()
        }
    }

    @Nested
    inner class IteratorTest {
        @Test
        fun `should iterate over elements`() {
            // given
            val givenBytes = byteArrayOf(1, 2, 3)
            val actualBytes = ByteArray(givenBytes.size)
            val iterator = bytesOf(givenBytes).iterator()

            // when
            var index = 0
            while (iterator.hasNext()) {
                actualBytes[index++] = iterator.next()
            }

            // then
            expectThat(actualBytes) isEqualTo givenBytes
        }

        @Test
        fun `should throw NoSuchElementException on next when there is no more element`() {
            // given
            val givenBytes = byteArrayOf(1, 2, 3)
            val iterator = bytesOf(givenBytes).iterator()
            while (iterator.hasNext()) {
                iterator.next()
            }

            // when
            val exception = tryCatching { iterator.next() }.exceptionOrNull()

            // then
            expectThat(exception).isA<NoSuchElementException>()
        }
    }
}
